File: server/fastapi/config/lifecycles/24_compute_functions.lifecycle.py
"""Slot 24 — discover server/computed_functions/*.compute.py and register into a
ComputeRegistry placed at app.state.compute_registry. Slot 26 reads it.

Multi-root, signature/scope validation, structured ledger — see fastify twin
24_compute_functions.lifecycle.mjs for the full contract.
"""
from __future__ import annotations

import importlib.util
import logging
import os
import re
from pathlib import Path
from typing import Any

from runtime_template_resolver import ComputeRegistry, ComputeScope

log = logging.getLogger("fastapi_server.compute_functions")

_VALID_SCOPES = {"STARTUP", "REQUEST"}
_NAME_RE = re.compile(r"^\d+_")


def _fallback_root() -> Path:
    fixtures = os.environ.get("APP_YAML_FIXTURES_DIR")
    if fixtures:
        return Path(fixtures).resolve().parent / "computed_functions"
    return Path(__file__).resolve().parents[3] / "computed_functions"


def _pick_roots() -> list[Path]:
    env = os.environ.get("COMPUTE_FUNCTIONS_DIRS")
    if env:
        return [Path(p) for p in env.split(":") if p]
    return [_fallback_root()]


def _name_from_filename(p: Path) -> str:
    return _NAME_RE.sub("", p.name).removesuffix(".compute.py")


def _load_module(path: Path):
    spec = importlib.util.spec_from_file_location(path.stem, path)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


def on_init(app, _config) -> None:
    app.state.compute_registry = None
    app.state.compute_registry_ledger = None


async def on_startup(app, _config) -> None:
    ledger: dict[str, Any] = {
        "event": "compute_functions_registered",
        "count": 0,
        "roots": [],
        "registered": [],
        "skipped": [],
        "collisions": [],
    }
    registry = ComputeRegistry()
    seen: dict[str, str] = {}

    for root in _pick_roots():
        if not root.is_dir():
            log.warning("compute_root_missing path=%s", root)
            continue
        ledger["roots"].append(str(root))
        for file in sorted(root.glob("*.compute.py")):
            name = _name_from_filename(file)
            try:
                mod = _load_module(file)
            except Exception as err:  # noqa: BLE001
                ledger["skipped"].append({"file": str(file), "reason": f"import failed: {err}"})
                continue
            fn = getattr(mod, "compute", None)
            if not callable(fn):
                ledger["skipped"].append({"file": str(file), "reason": f"compute() not callable (got {type(fn).__name__})"})
                continue
            scope = getattr(mod, "scope", None)
            if scope is None:
                log.info("compute_function_scope_default file=%s default=REQUEST", file)
                scope = "REQUEST"
            elif scope not in _VALID_SCOPES:
                ledger["skipped"].append({"file": str(file), "reason": f"invalid scope '{scope}' (expected STARTUP|REQUEST)"})
                continue
            if name in seen:
                ledger["collisions"].append({"name": name, "previous_source": seen[name], "new_source": str(file)})
                log.warning("compute_function_collision name=%s previous=%s new=%s", name, seen[name], file)
            seen[name] = str(file)
            cs = ComputeScope.STARTUP if scope == "STARTUP" else ComputeScope.REQUEST
            registry.register(name, fn, cs)
            ledger["registered"] = [e for e in ledger["registered"] if e["name"] != name]
            ledger["registered"].append({"name": name, "scope": scope, "source": str(file)})

    ledger["count"] = len(ledger["registered"])

    if ledger["count"] == 0 and os.environ.get("COMPUTE_FUNCTIONS_REQUIRE_AT_LEAST_ONE") == "true":
        raise RuntimeError("compute_functions: zero functions registered and COMPUTE_FUNCTIONS_REQUIRE_AT_LEAST_ONE=true")

    app.state.compute_registry = registry
    app.state.compute_registry_ledger = ledger
    log.info("compute-functions registered: %d (%s)", ledger["count"], ", ".join(f"{e['name']} ({e['scope']})" for e in ledger["registered"]))


async def on_shutdown(app, _config) -> None:
    app.state.compute_registry = None
    app.state.compute_registry_ledger = None

File: server/fastify/config/lifecycles/24_compute_functions.lifecycle.mjs
// Slot 24 — discover server/computed_functions/*.compute.mjs and register each into a
// ComputeRegistry decorated as `compute_registry`. Slot 26 reads it.
//
// Multi-root: COMPUTE_FUNCTIONS_DIRS env (colon-separated, like PATH). Defaults to
// $APP_YAML_FIXTURES_DIR/../computed_functions if set, else import.meta.url-relative.
// Last-wins on name collision; structured WARN per collision.
// Validation: signature (must be callable) + scope (STARTUP|REQUEST). Bad files skip
// with reason recorded in the ledger; boot does NOT abort unless
// COMPUTE_FUNCTIONS_REQUIRE_AT_LEAST_ONE=true and zero functions register.
import fp from "fastify-plugin";
import { ComputeRegistry, ComputeScope } from "@ployglot/runtime-template-resolver";
import { existsSync, readdirSync, statSync } from "node:fs";
import { resolve, dirname, basename } from "node:path";
import { pathToFileURL, fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));
const FALLBACK_ROOT = resolve(
  process.env.APP_YAML_FIXTURES_DIR
    ? resolve(process.env.APP_YAML_FIXTURES_DIR, "..", "computed_functions")
    : resolve(HERE, "../../../computed_functions")
);

const VALID_SCOPES = new Set(["STARTUP", "REQUEST"]);

function pickRoots() {
  const env = process.env.COMPUTE_FUNCTIONS_DIRS;
  if (env && env.length) return env.split(":").filter(Boolean);
  return [FALLBACK_ROOT];
}

function nameFromFilename(file) {
  return basename(file).replace(/^\d+_/, "").replace(/\.compute\.mjs$/, "");
}

async function plugin(fastify, _opts) {
  const ledger = {
    event: "compute_functions_registered",
    count: 0,
    roots: [],
    registered: [],
    skipped: [],
    collisions: [],
  };
  const registry = new ComputeRegistry();
  const seen = new Map(); // name → source path (for collision detection)

  for (const root of pickRoots()) {
    if (!existsSync(root) || !statSync(root).isDirectory()) {
      fastify.log.warn({ event: "compute_root_missing", path: root });
      continue;
    }
    ledger.roots.push(root);
    const files = readdirSync(root).filter((f) => f.endsWith(".compute.mjs")).sort();
    for (const f of files) {
      const file = resolve(root, f);
      const name = nameFromFilename(file);
      let mod;
      try {
        mod = await import(pathToFileURL(file).href);
      } catch (err) {
        ledger.skipped.push({ file, reason: `import failed: ${err.message}` });
        continue;
      }
      if (typeof mod.default !== "function") {
        ledger.skipped.push({ file, reason: `not a callable (typeof default = ${typeof mod.default})` });
        continue;
      }
      let scope = mod.scope;
      if (scope === undefined) {
        fastify.log.info({ event: "compute_function_scope_default", file, default: "REQUEST" });
        scope = "REQUEST";
      } else if (!VALID_SCOPES.has(scope)) {
        ledger.skipped.push({ file, reason: `invalid scope '${scope}' (expected STARTUP|REQUEST)` });
        continue;
      }
      if (seen.has(name)) {
        ledger.collisions.push({ name, previous_source: seen.get(name), new_source: file });
        fastify.log.warn({ event: "compute_function_collision", name, previous_source: seen.get(name), new_source: file });
      }
      seen.set(name, file);
      registry.register(name, mod.default, scope === "STARTUP" ? ComputeScope.STARTUP : ComputeScope.REQUEST);
      // Replace any earlier ledger entry for this name (last-wins).
      ledger.registered = ledger.registered.filter((e) => e.name !== name);
      ledger.registered.push({ name, scope, source: file });
    }
  }
  ledger.count = ledger.registered.length;

  if (ledger.count === 0 && process.env.COMPUTE_FUNCTIONS_REQUIRE_AT_LEAST_ONE === "true") {
    throw new Error("compute_functions: zero functions registered and COMPUTE_FUNCTIONS_REQUIRE_AT_LEAST_ONE=true");
  }

  fastify.decorate("compute_registry", registry);
  fastify.decorate("compute_registry_ledger", ledger);
  fastify.log.info(ledger);
}

const wrapped = fp(plugin, { name: "compute-functions", fastify: ">=4" });

export default async function lifecycle(server, _config) {
  await server.register(wrapped);
}
